perf/uncore_libpfm.go

```diff
@@ -246,147 +322,187 @@ func (c *uncoreCollector) Destroy() {
 	klog.V(5).Info("Attempting to update uncore perf_event stats")
 
-	for name, pmus := range c.cpuFiles {
-		for pmu, cpus := range pmus {
-			for cpu, file := range cpus {
-				stat, err := readPerfUncoreStat(file, name, cpu, pmu, c.topology)
+	for _, groupPMUs := range c.cpuFiles {
+		for pmu, group := range groupPMUs {
+			for cpu, file := range group.cpuFiles[group.leaderName] {
+				stat, err := readPerfUncoreStat(file, group, cpu, pmu, c.topology)
 				if err != nil {
-					return fmt.Errorf("unable to read from uncore perf_event_file (event: %q, CPU: %d, PMU: %s): %q", name, cpu, pmu, err.Error())
+					klog.Warningf("Unable to read from perf_event_file (event: %q, CPU: %d) for %q: %q", group.leaderName, cpu, pmu, err.Error())
+					continue
 				}
-				klog.V(5).Infof("Read uncore perf event (event: %q, CPU: %d, PMU: %s): %d", name, cpu, pmu, stat.Value)
 
-				stats.PerfUncoreStats = append(stats.PerfUncoreStats, *stat)
+				stats.PerfUncoreStats = append(stats.PerfUncoreStats, stat...)
 			}
 		}
 	}
 
 	return nil
 }
 
-func (c *uncoreCollector) setupRawNonGroupedUncore(event *CustomEvent, pmus []pmu) error {
-	klog.V(5).Infof("Setting up non-grouped raw perf uncore event %#v", event)
-
-	if event.Type == 0 {
-		// PMU isn't set. Register event for all PMUs.
-		for _, pmu := range pmus {
-			newEvent := CustomEvent{
-				Type:   pmu.typeOf,
-				Config: event.Config,
-				Name:   event.Name,
-			}
-			config := createPerfEventAttr(newEvent)
-			err := c.registerUncoreEvent(config, string(newEvent.Name), pmu.cpus, pmu.name)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
-	} else {
-		// Register event for the PMU.
-		config := createPerfEventAttr(*event)
-		pmu, err := getPMU(pmus, event.Type)
-		if err != nil {
-			return err
-		}
-		return c.registerUncoreEvent(config, string(event.Name), pmu.cpus, pmu.name)
+func (c *uncoreCollector) setupEvent(name string, pmus uncorePMUs, groupIndex int, leaderFileDescriptors map[string]map[uint32]int) error {
+	if !isLibpfmInitialized {
+		return fmt.Errorf("libpfm4 is not initialized, cannot proceed with setting perf events up")
 	}
-}
 
-func (c *uncoreCollector) setupNonGroupedUncore(name string, pmus []pmu) error {
-	perfEventAttr, err := getPerfEventAttr(name)
+	klog.V(5).Infof("Setting up grouped uncore perf event %s", name)
+
+	perfEventAttrMemory := C.malloc(C.ulong(unsafe.Sizeof(unix.PerfEventAttr{})))
+	defer C.free(perfEventAttrMemory)
+	err := readPerfEventAttr(name, perfEventAttrMemory)
 	if err != nil {
 		return err
 	}
-	defer C.free(unsafe.Pointer(perfEventAttr))
-
-	klog.V(5).Infof("Setting up non-grouped uncore perf event %s", name)
+	perfEventAttr := (*unix.PerfEventAttr)(perfEventAttrMemory)
 
 	// Register event for all memory controllers.
 	for _, pmu := range pmus {
 		perfEventAttr.Type = pmu.typeOf
-		err = c.registerUncoreEvent(perfEventAttr, name, pmu.cpus, pmu.name)
+		setGroupAttributes(perfEventAttr, leaderFileDescriptors[pmu.name][pmu.cpus[0]] == -1)
```

Comment:
`leaderFileDescriptors[pmu.name][pmu.cpus[0]] == -1` looks pretty complex to me. Do you have a test case that makes sure that it works? We are back to my comments for `collector`: there should be only one place to decide who the leader is.